javascript/ql/lib/semmle/javascript/dataflow/internal/CallGraphs.qll

```diff
@@ -279,6 +279,20 @@ module CallGraph {
     StepSummary::step(getAnAllocationSiteRef(node), result, objectWithMethodsStep())
   }
 
+  /**
+   * Holds if `function` flows to a property of `host` via non-local data flow.
+   */
+  pragma[nomagic]
+  private predicate complexMethodInstallation(
+    DataFlow::SourceNode host, DataFlow::FunctionNode function
+  ) {
+    not function = getAMethodOnObject(_) and
```

Comment:
```suggestion
    // To avoid double-recursion, we ignore any function that would be covered by 'Complex flow for the host object' in `impliedReceiverStep`
    not function = getAMethodOnObject(_) and
```

Reply:
Maybe I've misunderstood the purpose of `not function = getAMethodOnObject(_)` then.

My attempt at making sense of it was like this:

1. In `impliedReceiverStep` you've made it clear we want to split complex flow for host/function, to avoid double-recursion.
2. the `not function = getAMethodOnObject(_)` seems to ensure we don't do complex flow on host inside `complexMethodInstallation`, which is the predicate responsible for complex flow on functions.
3. therefore, `not function = getAMethodOnObject(_)` must help to achieve our goal of avoiding double-recursion.

but do let me know where my reasoning went wrong.

Reply:
The `not` is really just here to make the predicate smaller (and easier to inspect with quick eval). The predicate is called `complex` and the qldoc mentions that it excludes non-local data flow, so it's name and qldoc wouldn't really be right if they were included.

As for double recursion, that's a static property, like negative recursion, that can't be prevented by writing excluding tuples in the predicate body. `complexMethodInstallation` and `getAnAllocationSiteRef` are both part of the call graph recursive SCC. So if we write a conjunction like 
```codeql
complexMethodInstallation(host, function) and
pred = getAnAllocationSiteRef(host)
```
then that becomes a double-recursive join because we're joining two recursive predicates inside the recursion. (And perhaps I'm being overly cautious when wanting to avoid double recursion here, but that's a separate discussion)

Reply:
Thanks for talking this over offline :+1:

I can come up with a contrived example we wouldn't be able to handle due to the restriction. I don't know whether it's realistic to find this in any real code though :shrug: 

```js
function foo() {
    console.log("from foo", this.value);
}

var x = {};
x.value = 42;
x.foo = foo;

function fooFactoryFactory() {
    return function fooFactory() {
        return foo;
    }
}

var y = {}
y.value = 43;
y.foo = fooFactoryFactory()();
```

Overall, I can certainly see that the addition in this PR makes the analysis more powerful than it was before, so have approved the PR :+1: